### PR TITLE
C++: Fix issue with cpp/user-controlled-bypass

### PR DESCRIPTION
**Pull request by geoffw0**
Improve `cpp/user-controlled-bypass` to hopefully fix the `JVM_OUT_OF_MEM` errors we've seen on `hejiann/beautify` (likely affecting many other projects).  I have combined the large number of `matches` calls in this query into a much smaller number of `regexpMatch` calls, trusting the regexp compiler to compute them efficiently.

Note that I couldn't reproduce the OOM locally, but:
 - performance on this project is greatly improved (locally 97s -> 14s).  Performance and OOM issues tend to be closely related.
 - similar / more extreme improvement with reduced memory (8,000MB / 6,000MB).
 - performance on other projects I tried appears to be unaffected.
 - DCA should confirm.

I've also added extra tests because various edge cases weren't covered.

### CODE REVIEW

**Review by jketema — DISMISSED** [review]
LGTM provided DCA is happy.

File: cpp/ql/src/Security/CWE/CWE-290/AuthenticationBypass.ql [hunk 1]
```diff
@@ -15,76 +15,24 @@
...
-    s.matches("\"%.my\"") or
-    s.matches("\"%.hk\"") or
-    s.matches("\"%.lt\"") or
-    s.matches("\"%.cc\"") or
-    s.matches("\"%.sg\"") or
-    s.matches("\"%.io\"") or
-    s.matches("\"%.edu\"") or
-    s.matches("\"%.gov\"")
+    s.regexpMatch("\"(www\\.|http:|https:).*\"") or
+    s.regexpMatch("\".*\\.(" + concat(getATopLevelDomain(), "|") + ")\"")
```
Comment (MathiasVP) [inline cpp/ql/src/Security/CWE/CWE-290/AuthenticationBypass.ql:35]: The compiler may be able to infer that `getATopLevelDomain()` is non-empty, but in case it can't we should probably use `strictconcat`.
```suggestion
    s.regexpMatch("\".*\\.(" + strictconcat(getATopLevelDomain(), "|") + ")\"")
```

**Comment by geoffw0** [comment]
DCA LGTM.

**Review by MathiasVP — APPROVED** [review]
